Auto update check

optizone opened the pull request:
Provides auto update check feature. 
Also improves way of manual update check. 

On start Cutter checks current available version. In case of available update, shows this window
(Instead of "Some newer version", there will be newest version number)
![image](https://user-images.githubusercontent.com/42874998/53681422-7beb0080-3cfa-11e9-9a82-bbc80807c5f0.png)




P.S. Prints `QNetworkReplyHttpImplPrivate::_q_startOperation was called more than once QUrl("https://api.github.com/repos/radareorg/cutter/releases/latest")` in console when checks.
It's a [known](https://bugreports.qt.io/browse/QTBUG-72463) bug. Fixed in 
5.12.2 Qt version.

ITAYC0HEN commented:
Thank you :)
We don't want Cutter to open connection to the internet by default. This feature should only be triggered by the user in these cases:
1. Explicitly clicking a button saying "Check for available updates"
2. Opt-in to the "check for new update on start"

The opt in should be available both in the Configuration menu as well as in the Welcome dialog.

Please check here for more information: 
https://github.com/radareorg/cutter/issues/1177


Now, please follow our template for Pull Requests and describe thoroughly the changes you made. What does it add to the current "Update Button" we already have, etc

After enough information and test plan we would be able to conduct a code review :)

optizone commented:
Now Cutter's users need to go to About page to check for available updates. So, if Cutter did it itself any time one starts, there wouldn't be situation when there was a new version of Cutter but users didn't know about it.

Changes:
- Added checkboxes to About page and to Welcome dialog, which allow user to choose whether he wants to use auto update check
- Added `VersionChecker` class encapsulating version check process and remove one from `AboutDialog` class

Tests:
- Run check without internet connection
- Run check with bad link
- Run a bunch of checks in a row
- Run check with 1 ms timeout
- Create a bunch of `VersionChecker` instances and run one check per one instance
- Run check normally

Demonstration:
- Screenshot of message in case of new version being attached in first comment
- Workflow of "Check for updates" button is still

![image](https://user-images.githubusercontent.com/42874998/53683751-8cf53b00-3d15-11e9-86f3-b801ec0d8f5f.png)

![image](https://user-images.githubusercontent.com/42874998/53683753-92eb1c00-3d15-11e9-8cf0-6d0d482218f5.png)

Welcome dialog:
![image](https://user-images.githubusercontent.com/42874998/53683858-05a8c700-3d17-11e9-976e-29a5b06538d0.png)

ITAYC0HEN commented on src/common/VersionChecker.h:
```diff
@@ -0,0 +1,55 @@
+      \fn void VersionChecker::checkCurrentVersion(time_t timeoutMs)
+
+      Sends request to determine current version of Cutter.
+      If there are no respond in \a timeoutMs milliseconds, emits
```
please change "there are no respond" to "there is no response"

ITAYC0HEN commented on src/common/VersionChecker.h:
```diff
@@ -0,0 +1,55 @@
+    /*!
+      \fn VersionChecker::checkComplete(const QString& verson, const QString& errorMsg)
+
+      The signal is emited when check has been done with empty \a errorMsg string.
```
"emited" -> "emitted"

"with empty" -> "with an empty"

ITAYC0HEN commented on src/common/VersionChecker.h:
```diff
@@ -0,0 +1,55 @@
+      \fn VersionChecker::checkComplete(const QString& verson, const QString& errorMsg)
+
+      The signal is emited when check has been done with empty \a errorMsg string.
+      In case of error \a currVerson is empty and \a errorMsg contains description
```
"In case of error \a currVerson" - >
"In case of an error, \a currVerson"

ITAYC0HEN commented on src/dialogs/AboutDialog.cpp:
```diff
@@ -86,56 +89,34 @@ void AboutDialog::on_checkForUpdatesButton_clicked()
+            QMessageBox mb;
+            mb.setStandardButtons(QMessageBox::Ok);
+            mb.setWindowTitle(tr("Version control"));
+            mb.setIcon(QMessageBox::Information);
```
You probably can save some lines if you'll use [this](https://doc.qt.io/qt-5/qmessagebox.html#QMessageBox-1) constructor for QMessageBox

ITAYC0HEN commented on src/dialogs/AboutDialog.cpp:
```diff
@@ -75,9 +80,7 @@ void AboutDialog::on_showPluginsButton_clicked()
 
 void AboutDialog::on_checkForUpdatesButton_clicked()
```
please document this function is you did **very well** with others. \brief, etc

ITAYC0HEN commented on src/dialogs/AboutDialog.cpp:
```diff
@@ -86,56 +89,34 @@ void AboutDialog::on_checkForUpdatesButton_clicked()
 }
 
-void AboutDialog::serveVersionCheckReply(QNetworkReply *reply)
+void AboutDialog::on_updatesCheckBox_stateChanged(int state)
```
please document this function is you did **very well** with others. \brief, etc

ITAYC0HEN commented on src/dialogs/AboutDialog.cpp:
```diff
@@ -86,56 +89,29 @@ void AboutDialog::on_checkForUpdatesButton_clicked()
+            QMessageBox::critical(nullptr, tr("Error!"), error);
+        } else {
+            QString text = (version == CUTTER_VERSION_FULL)
+                           ? tr("You have latest version and no need to update!")
```
"Cutter is up to date."

a1ext commented:
Thanks for the PR.

I got strange behavior, I did:
- clicked `About` -> `Check for updates`
- the two message boxes appeared one after another:
![image](https://user-images.githubusercontent.com/14978853/53696515-aad7a400-3dd8-11e9-8ef6-02e2b4b7d3f9.png)

a1ext commented:
It seems like your timer got expired (you show the first message box with an error) and then the network request got finished and you shown that version is ok.

a1ext commented:
One more thing, when `Check for updates on start` is turned on, then cutter appears like the following until network request is finished:

![image](https://user-images.githubusercontent.com/14978853/53696647-300f8880-3dda-11e9-9e3c-fd574931b218.png)

optizone commented:
It really seems like `QNetworkAccessManager::get(QNetworkRequest)` blocks main thread, but it's not expected behavior. Maybe on windows it works like that. At least I tested only on Linux.

optizone commented:
@a1ext may I ask you check behavior with this patch? I don't have access to windows until tomorrow and I'm very curious why it behaves this way.

a1ext commented:
> @a1ext may I ask you check behavior with this patch? I don't have access to windows until tomorrow and I'm very curious why it behaves this way.

Yeah, sure. Which patch did you mean? This commit b735134?

optizone commented:
> > @a1ext may I ask you check behavior with this patch? I don't have access to windows until tomorrow and I'm very curious why it behaves this way.
> 
> Yeah, sure. Which patch did you mean? This commit [b735134](https://github.com/radareorg/cutter/commit/b7351346d2956a2fdea7dd875e0853044a13f155)?

yes

a1ext commented:
> > > @a1ext may I ask you check behavior with this patch? I don't have access to windows until tomorrow and I'm very curious why it behaves this way.
> > 
> > 
> > Yeah, sure. Which patch did you mean? This commit [b735134](https://github.com/radareorg/cutter/commit/b7351346d2956a2fdea7dd875e0853044a13f155)?
> 
> yes

No, this doesn't fix the problem.

I think it has to be implemented according to the following:
- network communication should be made in a separate thread
- the auto-check itself has to be started after the binary is chosen (e.g. after main window appears)

@ITAYC0HEN what do you think?

ITAYC0HEN commented:
> network communication should be made in a separate thread

Agree - **must** be running in the background to not stop the user from working

> the auto-check itself has to be started after the binary is chosen (e.g. after main window appears)

I disagree, this check should take 2 seconds for a normal computer, and it should show when Cutter is executing. Otherwise, people would waste 5 minutes analyzing a big binary and just after they finish they will see the notification about a new version. 
They'd be happy to avoid this and update Cutter before starting the heavy analysis

a1ext commented:
> > network communication should be made in a separate thread
> 
> Agree - **must** be running in the background to not stop the user from working
> 
> > the auto-check itself has to be started after the binary is chosen (e.g. after main window appears)
> 
> I disagree, this check should take 2 seconds for a normal computer, and it should show when Cutter is executing. Otherwise, people would waste 5 minutes analyzing a big binary and just after they finish they will see the notification about a new version.
> They'd be happy to avoid this and update Cutter before starting the heavy analysis

I didn't say about 5 minutes. I mean check has to run when MainWindow is going to be shown, it will be finished right after all the gadgets are filled with data.

ITAYC0HEN commented:
I understood. And I have have to disagree.

Imagine some John Doe, which is a daily user of Cutter, he updates Cutter to the new release immediately when possible! :D 
Now our John open Cutter because he wants to analyze some firmware of a printer machine. John opens Cutter, selects a file, selects the default analysis and then waits 25 minutes for Cutter to finish the analysis. Then, the MainWindow is opened, the widgets are filled with data and then BOOOM. MsgBox saying "You have a new update available! It is your lucky day <3".

Now John will download the new version and would have to do everything again and to wait again 25 minutes for the new Cutter to analyze the firmware.


Instead, if the Auto Update check had happened at the start, our John was happy because he just opened Cutter and then he sees the "lucky day!" message. He will quickly download the new version and will analyze his firmware with the new Cutter.

a1ext commented:
> I understood. And I have have to disagree.
> 
> Imagine some John Doe, which is a daily user of Cutter, he updates Cutter to the new release immediately when possible! :D
> Now our John open Cutter because he wants to analyze some firmware of a printer machine. John opens Cutter, selects a file, selects the default analysis and then waits 25 minutes for Cutter to finish the analysis. Then, the MainWindow is opened, the widgets are filled with data and then BOOOM. MsgBox saying "You have a new update available! It is your lucky day <3".
> 
> Now John will download the new version and would have to do everything again and to wait again 25 minutes for the new Cutter to analyze the firmware.
> 
> Instead, if the Auto Update check had happened at the start, our John was happy because he just opened Cutter and then he sees the "lucky day!" message. He will quickly download the new version and will analyze his firmware with the new Cutter.



> I understood. And I have have to disagree.
> 
> Imagine some John Doe, which is a daily user of Cutter, he updates Cutter to the new release immediately when possible! :D
> Now our John open Cutter because he wants to analyze some firmware of a printer machine. John opens Cutter, selects a file, selects the default analysis and then waits 25 minutes for Cutter to finish the analysis. Then, the MainWindow is opened, the widgets are filled with data and then BOOOM. MsgBox saying "You have a new update available! It is your lucky day <3".
> 
> Now John will download the new version and would have to do everything again and to wait again 25 minutes for the new Cutter to analyze the firmware.
> 
> Instead, if the Auto Update check had happened at the start, our John was happy because he just opened Cutter and then he sees the "lucky day!" message. He will quickly download the new version and will analyze his firmware with the new Cutter.

lol) I got your point, ok then

optizone commented:
Ok. There are few things I want to say. 
First of all, `QNetworkAccessManager` works asynchronously by default, so there is no need to move it to new thread. This situation on windows **might** be because of other causes. 
Second of all, by now I can not repeat behavior @a1ext has shown. But it's because test program on windows wont even connect to the Internet. Now I investigate this issue with my windows env.

ITAYC0HEN commented:
Thank you, we'll wait for further update. We won't merge until the windows
issue is fixed. Also, id like to remind that you can download Windows
artifacts from our appveyor build system :)

On Mon, Mar 4, 2019, 20:50 optizone <notifications@github.com> wrote:

> Ok. There are few things I want to say.
> First of all, QNetworkAccessManager works asynchronously by default, so
> there is no need to move it to new thread. This situation on windows
> *might* be because of other causes.
> Second of all, by now I can not repeat behavior @a1ext
> <https://github.com/a1ext> has shown. But it's because test program on
> windows wont even connect to the Internet. Now I investigate this issue
> with my windows env.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/radareorg/cutter/pull/1235#issuecomment-469374551>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATP2cptBaIE2jhxIst4V82vy0UbS4DYTks5vTWr_gaJpZM4baV21>
> .
>

ITAYC0HEN commented on src/Main.cpp:
```diff
@@ -34,6 +37,32 @@ int main(int argc, char *argv[])
+        [](const QString & version, const QString & error) {
+            if (error == "" && version != CUTTER_VERSION_FULL) {
+                QMessageBox mb;
+                mb.setWindowTitle(QObject::tr("Version control"));
```
I would be happy to see the following MsgBox:
```
There's an update available for Cutter.
Current version: XX
Latest version: YY
[ Download ] [ Don't check for updates ]  [ Cancel ]
```

Clicking "Download" will open the release link

ITAYC0HEN commented on src/Main.cpp:
```diff
@@ -34,6 +37,32 @@ int main(int argc, char *argv[])
+                mb.setText("<b>" + QObject::tr("Current version:") + "</b> " CUTTER_VERSION_FULL "<br/>"
+                           + "<b>" + QObject::tr("Latest version:") + "</b> " + version + "<br/><br/>"
+                           + QObject::tr("For update, please check the link:<br/>")
+                           + "<a href=\"https://github.com/radareorg/cutter/releases\">"
```
If possible, I would take the user immediately to the specific release page and not to the generic page

a1ext replied:
Yeah, there is `html_url` in the response, it's exactly what you need.

optizone replied:
I think we can download directly from Cutter, so user don't need to do it himself.

ITAYC0HEN replied:
No, meantime let the user choose which binary of the release (32b, 64b,linux,windows, etc)

optizone replied:
But we know what binaries user uses, so we can help him aviod going to download page and doing all this stuff. Maybe we should give him possibility to go to this page, but button "Download" will download binaries? I did it in last commits.

optizone commented:
I have some news. This bug appears only on Windows and Android. Code works correctly and in different thread. UI freezes appears because when first request is send Windows loads some DLL's and this blocks main thread (on my computer it is 1 second +- 60 ms and I think timings are similar on other machines). After first request is send next ones works fine. So, I see only four options:
- Load this libs before UI is launched, so user won't see freeze
- Leave it as it is (not as good as first option)
- Send request while analysis is going so UI freeze is not very noticeable 
- Do not provide this feature (not cool either)

optizone commented:
So, I tested first option from my last comment. And I'd say that it works pretty fine. DLL load delay might even become less long (maybe compiler optimized something) and I think now windows issue is solved.

optizone commented:
Btw, now Cutter will show "version control" dialog if you have latest version. It's done for debug purposes.

a1ext commented:
Now its better. It would be nice if `Download` button will use save file name dialog, because we know an archive name. It will allow user to rename it before save

optizone commented:
So, don't you mind this hack for Windows?

ITAYC0HEN commented:
sorry I had a review ready in the last few days which I didn't submit by accident
